### ISSUE TITLE
ClassNotFoundException: androidx.compose.foundation.layout.InnerPadding

### ISSUE THREAD

**Issue by peterfortuin** [airbnb/Showkase#111]
This project looks really promising, but I run into problems when I try to integrate this into my application.

As soon as I run
`startActivity(createShowkaseBrowserIntent(this))`
the application crashes with the following stacktrace:

```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.peerke.amuza.debug, PID: 26801
    java.lang.NoClassDefFoundError: Failed resolution of: Landroidx/compose/foundation/layout/InnerPadding;
        at com.airbnb.android.showkase.ui.ShowkaseBrowserAppKt$ShowkaseBrowserApp$2.invoke(Unknown Source:1)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(ComposableLambda.kt:151)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(Unknown Source:10)
        at androidx.compose.material.ScaffoldKt$ScaffoldContent$1.invoke(Scaffold.kt:330)
        at androidx.compose.material.ScaffoldKt$ScaffoldContent$1.invoke(Unknown Source:10)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(ComposableLambda.kt:142)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(Unknown Source:10)
        at androidx.compose.material.ScaffoldKt$ScaffoldSlot$1.invoke(Scaffold.kt:388)
        at androidx.compose.material.ScaffoldKt$ScaffoldSlot$1.invoke(Unknown Source:13)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(ComposableLambda.kt:151)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(Unknown Source:10)
        at androidx.compose.foundation.layout.StackKt$Stack$stackChildren$1.invoke(Stack.kt:54)
        at androidx.compose.foundation.layout.StackKt$Stack$stackChildren$1.invoke(Unknown Source:10)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(ComposableLambda.kt:142)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(Unknown Source:10)
        at androidx.compose.ui.LayoutKt.Layout(Layout.kt:675)
        at androidx.compose.foundation.layout.StackKt.Stack(Stack.kt:56)
        at androidx.compose.material.ScaffoldKt.ScaffoldSlot(Scaffold.kt:388)
        at androidx.compose.material.ScaffoldKt.ScaffoldContent(Scaffold.kt:325)
        at androidx.compose.material.ScaffoldKt.access$ScaffoldContent(Unknown Source:0)
        at androidx.compose.material.ScaffoldKt$Scaffold$child$1$1$1$1.invoke(Scaffold.kt:210)
        at androidx.compose.material.ScaffoldKt$Scaffold$child$1$1$1$1.invoke(Unknown Source:13)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(ComposableLambda.kt:151)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(Unknown Source:10)
        at androidx.compose.foundation.layout.StackKt$Stack$stackChildren$1.invoke(Stack.kt:54)
        at androidx.compose.foundation.layout.StackKt$Stack$stackChildren$1.invoke(Unknown Source:10)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(ComposableLambda.kt:142)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(Unknown Source:10)
        at androidx.compose.ui.LayoutKt.Layout(Layout.kt:675)
        at androidx.compose.foundation.layout.StackKt.Stack(Stack.kt:56)
        at androidx.compose.material.ScaffoldKt$Scaffold$child$1$1.invoke(Scaffold.kt:209)
        at androidx.compose.material.ScaffoldKt$Scaffold$child$1$1.invoke(Unknown Source:10)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(ComposableLambda.kt:142)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(Unknown Source:10)
        at androidx.compose.runtime.AmbientKt.Providers(Ambient.kt:175)
        at androidx.compose.material.SurfaceKt$Surface$1.invoke(Surface.kt:107)
        at androidx.compose.material.SurfaceKt$Surface$1.invoke(Unknown Source:10)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(ComposableLambda.kt:142)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(Unknown Source:10)
        at androidx.compose.ui.LayoutKt.Layout(Layout.kt:675)
        at androidx.compose.material.SurfaceKt.SurfaceLayout(Surface.kt:129)
        at androidx.compose.material.SurfaceKt.Surface-biUpMIw(Surface.kt:97)
        at androidx.compose.material.ScaffoldKt$Scaffold$child$1.invoke(Scaffold.kt:204)
        at androidx.compose.material.ScaffoldKt$Scaffold$child$1.invoke(Unknown Source:13)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(ComposableLambda.kt:151)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(Unknown Source:10)
        at androidx.compose.material.ScaffoldKt.Scaffold-tiGujGI(Scaffold.kt:240)
E/AndroidRuntime:     at com.airbnb.android.showkase.ui.ShowkaseBrowserAppKt.ShowkaseBrowserApp(ShowkaseBrowserApp.kt:40)
        at com.airbnb.android.showkase.ui.ShowkaseBrowserActivity$onCreate$1.invoke(ShowkaseBrowserActivity.kt:37)
        at com.airbnb.android.showkase.ui.ShowkaseBrowserActivity$onCreate$1.invoke(Unknown Source:10)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(ComposableLambda.kt:142)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(Unknown Source:10)
        at androidx.compose.ui.selection.SelectionContainerKt$SelectionContainer$3$1.invoke(SelectionContainer.kt:88)
        at androidx.compose.ui.selection.SelectionContainerKt$SelectionContainer$3$1.invoke(Unknown Source:10)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(ComposableLambda.kt:142)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(Unknown Source:10)
        at androidx.compose.ui.LayoutKt.Layout(Layout.kt:675)
        at androidx.compose.ui.selection.SelectionLayoutKt.SelectionLayout(SelectionLayout.kt:80)
        at androidx.compose.ui.selection.SelectionContainerKt$SelectionContainer$3.invoke(SelectionContainer.kt:85)
        at androidx.compose.ui.selection.SelectionContainerKt$SelectionContainer$3.invoke(Unknown Source:10)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(ComposableLambda.kt:142)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(Unknown Source:10)
        at androidx.compose.runtime.AmbientKt.Providers(Ambient.kt:175)
        at androidx.compose.ui.selection.SelectionContainerKt.SelectionContainer(SelectionContainer.kt:82)
        at androidx.compose.ui.selection.SelectionContainerKt.SelectionContainer(SelectionContainer.kt:39)
        at androidx.compose.ui.platform.WrappedComposition$setContent$1$1$1$1.invoke(Wrapper.kt:274)
        at androidx.compose.ui.platform.WrappedComposition$setContent$1$1$1$1.invoke(Unknown Source:10)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(ComposableLambda.kt:142)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(Unknown Source:10)
        at androidx.compose.runtime.AmbientKt.Providers(Ambient.kt:175)
        at androidx.compose.ui.platform.AmbientsKt.ProvideCommonAmbients(Ambients.kt:109)
        at androidx.compose.ui.platform.AndroidAmbientsKt$ProvideAndroidAmbients$3.invoke(AndroidAmbients.kt:101)
        at androidx.compose.ui.platform.AndroidAmbientsKt$ProvideAndroidAmbients$3.invoke(Unknown Source:10)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(ComposableLambda.kt:142)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(Unknown Source:10)
        at androidx.compose.runtime.AmbientKt.Providers(Ambient.kt:175)
        at androidx.compose.ui.platform.AndroidAmbientsKt.ProvideAndroidAmbients(AndroidAmbients.kt:93)
        at androidx.compose.ui.platform.WrappedComposition$setContent$1$1$1.invoke(Wrapper.kt:273)
        at androidx.compose.ui.platform.WrappedComposition$setContent$1$1$1.invoke(Unknown Source:10)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(ComposableLambda.kt:142)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(Unknown Source:10)
        at androidx.compose.runtime.AmbientKt.Providers(Ambient.kt:175)
        at androidx.compose.ui.platform.WrappedComposition$setContent$1$1.invoke(Wrapper.kt:272)
        at androidx.compose.ui.platform.WrappedComposition$setContent$1$1.invoke(Unknown Source:10)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(ComposableLambda.kt:142)
        at androidx.compose.runtime.internal.ComposableLambda.invoke(Unknown Source:10)
        at androidx.compose.runtime.ComposerKt.invokeComposable(Composer.kt:2686)
        at androidx.compose.runtime.Composer.composeInitial(Composer.kt:1853)
        at androidx.compose.runtime.Recomposer.composeInitial$runtime_release(Recomposer.kt:241)
        at androidx.compose.runtime.CompositionImpl.setContent(Composition.kt:110)
        at androidx.compose.ui.platform.WrappedComposition$setContent$1.invoke(Wrapper.kt:264)
        at androidx.compose.ui.platform.WrappedComposition$setContent$1.invoke(Unknown Source:3)
E/AndroidRuntime:     at androidx.compose.ui.platform.AndroidComposeView.onAttachedToWindow(AndroidComposeView.kt:557)
        at android.view.View.dispatchAttachedToWindow(View.java:19701)
        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3455)
        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3462)
        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3462)
        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3462)
        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3462)
        at android.view.ViewGroup.dispatchAttachedToWindow(ViewGroup.java:3462)
        at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2260)
        at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1940)
        at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:8041)
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1163)
        at android.view.Choreographer.doCallbacks(Choreographer.java:986)
        at android.view.Choreographer.doFrame(Choreographer.java:902)
        at android.view.Choreographer$FrameHandler.handleMessage(Choreographer.java:1091)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7710)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:516)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
     Caused by: java.lang.ClassNotFoundException: Didn't find class "androidx.compose.foundation.layout.InnerPadding" on path: DexPathList[[zip file "/data/app/com.peerke.amuza.debug-2piQ_OS2K27Bh1BbnrT3cg==/base.apk"],nativeLibraryDirectories=[/data/app/com.peerke.amuza.debug-2piQ_OS2K27Bh1BbnrT3cg==/lib/arm64, /data/app/com.peerke.amuza.debug-2piQ_OS2K27Bh1BbnrT3cg==/base.apk!/lib/arm64-v8a, /system/lib64, /system/product/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:196)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
        	... 113 more

```

I use the following versions:
Showkase: 1.0.0-alpha02
Gradle Wrapper: 6.6.1
Android Gradle Plugin: 4.1.0-rc03
Jetpack Compose: 1.0.0-alpha03

**Comment by vinaygaba** [comment]
I somehow completely missed noticing this issue! Thanks for reporting it, I'm going to take a look!

**Comment by vinaygaba** [comment]
@peterfortuin I think this might be related to the difference between the compose versions used in the library and the one you are using. Can you try using the `alpha02` version of Jetpack Compose and giving it a shot?

I recently just bumped up the library to the latest version - https://github.com/airbnb/Showkase/pull/112. I'll be doing a release in the next couple days so you should be able to use the latest library version soon. In the meanwhile, I'm curious to see if it works fine when you use Jetpack Compose v `1.0.0-alpha02`

**Comment by vinaygaba** [comment]
@peterfortuin Can you try the latest release of this library. I believe that will certainly fix your problem!

```
implementation "com.airbnb.android:showkase:1.0.0-alpha03"
kapt "com.airbnb.android:showkase-processor:1.0.0-alpha03"
```

**Comment by peterfortuin** [comment]
If I downgrade to compose 1.0.0-alpha02 I have another problem. Maybe not related to this. I'm not sure if I can upgrade to the latest version of compose. Have to check.
To be continued.

**Comment by vinaygaba** [comment]
@peterfortuin You don't need to downgrade anymore! You can use the latest version of compose as this library has been upgraded!

**Comment by peterfortuin** [comment]
I have included `showkase:1.0.0-alpha03` into my project with `compose:1.0.0-alpha04` and that seems to work. 

But now that I have it working, I'm not sure how useful it is for me. 

Some things that I noticed:

- The Components part directly crashes for me, because I use ViewModels on a lot of places inside @Composable functions and those ViewModels have constructor parameters. Of course Showkase doesn't know what to do with those parameters of course.
This is not only a problem with Showkase, but this is a general problem I have with the @Preview annotation of Compose. 
- Currently I have the colors of my application as Color Resources in xml and read them with colorResource function into my Theme and other places. As far I can see from the documentation I have to add the colors again to show up in Showkaze. But I prefer not to write down the colors in two places. (Changes are big that they will not match any more after some changes in the future). Is there no way to read them from the resources like this?

```
@ShowkaseColor(name = "Primary Color", group = "Material Design")
val primaryColor = colorResource(R.color.colorPrimary)
```

- Same problem for the typography. I create those inside my AppTheme @Composable function. I copy them from the MaterialTheme and change some values. I can't do that outside of the function and annotate that. 

I hope you can do something with my feedback or maybe tell me how I can structure my application differently to be more compatible with Showkase.

**Comment by vinaygaba** [comment]
Hey @peterfortuin, 
I'm glad the original issue that you reported is fixed now. I can try to address some of the points you raised -

1. Showkase (and Jetpack Compose) does have a way to [provide parameters](https://developer.android.com/reference/kotlin/androidx/ui/tooling/preview/PreviewParameter) for preview. It does so using the `@PreviewParameter` annotation. I recently added support for it but I still have [another PR](https://github.com/airbnb/Showkase/pull/110) to go, hence I have not mentioned it in the README just yet. This is how you can use it to provide parameters and use view models with Showkase. I'm not sure if `@Preview` in Android Studio allows using ViewModel, but Showkase certainly does! Here is what it would roughly look like in practice. 

```kotlin

data class ViewModelParameters(
    val param1: String,
    val param2: Int
)


@ShowkaseComposable("Component with ViewModel")
@Composable
fun test(
    @PreviewParameter(provider = CustomPreviewProvider::class) viewModelParameters: ViewModelParameters
) {
    val factory = MyViewModelFactory(viewModelParameters.param1, viewModelParameters.param2)
    val viewModel: MyViewModel = viewModel(factory = factory)
    Text(viewModel.param1)
}

class MyViewModel(
    val param1: String,
    val param2: Int
): ViewModel() {}

class MyViewModelFactory(
    val param1: String,
    val param2: Int
): ViewModelProvider.Factory {
    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(MyViewModel::class.java)) {
            return MyViewModel(param1, param2) as T
        }
        throw IllegalArgumentException("Model class ${modelClass.name} not valid")
    }

}

class CustomPreviewProvider: PreviewParameterProvider<ViewModelParameters> {
    override val values: Sequence<ViewModelParameters>
        get() = sequenceOf(
            ViewModelParameters(param1 = "Param1", param2 = 5)
        )
}

```

2. This is a reasonable request but there is no easy way to solve this as `colorResource` can only be used in the scope of a Composable function and the code that is generated doesn't have @Composable scope. I can think of a few approaches to potentially support this but it's going to be a good amount of work. I might give this a shot at some point. 

I'll close this ticket for now since the original problem is fixed but I'll create a separate issue (2) as a feature request.
